lib/src/components/NxModal/NxModal.tsx

```diff
@@ -4,22 +4,42 @@
  * the terms of the Eclipse Public License 2.0 which accompanies this
  * distribution and is available at https://www.eclipse.org/legal/epl-2.0/.
  */
-import React, {FunctionComponent} from 'react';
+import React, {FunctionComponent, useEffect} from 'react';
 import classnames from 'classnames';
 
-import {Props, propTypes} from './types';
+import {CloseHandler, Props, propTypes} from './types';
 
 import './NxModal.scss';
 
-const NxModal: FunctionComponent<Props> = ({className, ...attrs}) => {
+const currentModalCloseHandlers: CloseHandler[] = [];
 
+const NxModal: FunctionComponent<Props> = ({className, onClose, ...attrs}) => {
   const modalClasses = classnames('nx-modal', className);
+
+  const modalCloseListener = ({ key }: { key: string}) => {
+    if (key === 'Escape' && currentModalCloseHandlers.length
+        && onClose === currentModalCloseHandlers[currentModalCloseHandlers.length - 1]) {
+      onClose();
+    }
+  };
+
+  const removeCloseHandlerListener = () => {
+    const idx = currentModalCloseHandlers.indexOf(onClose);
+    currentModalCloseHandlers.splice(idx, 1);
+    document.removeEventListener('keydown', modalCloseListener);
+  };
+
+  useEffect(function() {
+    document.addEventListener('keydown', modalCloseListener);
```

Comment:
Why the deviation from my suggestion where the listener was registered outside of the `NxModal` invocation?  This needlessly registers it one time per modal instead of one time globally

Reply:
> Why the deviation from my suggestion where the listener was registered outside of the `NxModal` invocation? This needlessly registers it one time per modal instead of one time globally

Discussing the approach with @raveneyex he mentioned that might introduce a memory leak since it's always registered. With the current approach once there are no modals there are no more calls to the validation logic either which seems cleaner

Reply:
> my suggestion where the listener was registered outside of the `NxModal` invocation

Who would clean up such a listener once all the modals are closed?

Reply:
> Who would clean up such a listener once all the modals are closed?

No one.  It is registered when the page first loads and it stays there until the browser tears everything down when the page is closed.  I am fine with this.

However, I do recognize that the ideal scenario would be to have zero listeners registered when there are no modals open, and exactly one listener registered when there are one or more modals open - a combination which neither of our solutions achieves.  I think it could be achieved fairly easily by putting some list-size-based conditions around the registering and unregistering of the listener in your impl.

Reply:
> I think it could be achieved fairly easily by putting some list-size-based conditions around the registering and unregistering of the listener in your impl.

Can you elaborate on this? I don't quite follow.

Reply:
> > I think it could be achieved fairly easily by putting some list-size-based conditions around the registering and unregistering of the listener in your impl.
> 
> Can you elaborate on this? I don't quite follow.

I think it refers to only registering the listener in the use effect if there is something on the closeHandler array and only deregistering it when there is nothing else on the array. I had tried something similar but couldn't get it to work at that time. I'm looking into the approach again. @rpokorny let me know if you had a different idea for this

Reply:
> > > I think it could be achieved fairly easily by putting some list-size-based conditions around the registering and unregistering of the listener in your impl.
> > 
> > 
> > Can you elaborate on this? I don't quite follow.
> 
> I think it refers to only registering the listener in the use effect if there is something on the closeHandler array and only deregistering it when there is nothing else on the array. I had tried something similar but couldn't get it to work at that time. I'm looking into the approach again. @rpokorny let me know if you had a different idea for this

Improved the listener registration so it's only registered once per modal stacking: 463660d